### PR TITLE
fix: Get sqlcmd utility file path from container instead of const file path

### LINKED ISSUE #1220: [Bug]: MsSql health check does not complete on newest container image
### Testcontainers version

3.9.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x64

### .NET version

8.0.303

### Docker version

```shell
Client:
 Version:           25.0.5
 API version:       1.44
 Go version:        go1.21.10
 Git commit:        d260a54c81efcc3f00fe67dee78c94b16c2f8692
 Built:             Sun May 12 07:25:43 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          25.0.5
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.10
  Git commit:       e63daec8672d77ac0b2b5c262ef525c7cf17fd20
  Built:            Sun May 12 07:25:43 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.10
  GitCommit:        4e1fe7492b9df85914c389d1f15a3ceedbb280ac
 runc:
  Version:          1.1.12
  GitCommit:        51d5e94601ceffbbd85688df1c928ecccbfa4685
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### Docker info

```shell
Client:
 Version:    25.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.0
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx

Server:
 Containers: 29
  Running: 5
  Paused: 0
  Stopped: 24
 Images: 15
 Server Version: 25.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 4e1fe7492b9df85914c389d1f15a3ceedbb280ac
 runc version: 51d5e94601ceffbbd85688df1c928ecccbfa4685
 init version:
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 5.15.153.1-microsoft-standard-WSL2
 Operating System: Rancher Desktop WSL Distribution
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 15.58GiB
 Name: CCD-0024
 ID: 398be532-db59-47b3-bcf7-d989f4f09517
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

### What happened?

When using the `MsSql` package with the newest container image [mcr.microsoft.com/mssql/server:2022-latest](https://mcr.microsoft.com/en-us/product/mssql/server/tags) with a digest of `sha256:c1aa8afe9b06eab64c9774a4802dcd032205d1be785b1fd51e1c0151e7586b74`, the [health check specified in the waiting strategy](https://github.com/testcontainers/testcontainers-dotnet/blob/develop/src/Testcontainers.MsSql/MsSqlBuilder.cs#L132-L145) never completes, even though the logs of the SQL server container show it being ready, leading to a timeout.

This behavior is not present when using a slightly older container image version, e.g. [mcr.microsoft.com/mssql/server:2022-CU13-ubuntu-22.04](https://mcr.microsoft.com/en-us/product/mssql/server/tags) with a digest of `sha256:c4369c38385eba011c10906dc8892425831275bb035d5ce69656da8e29de50d8`.

### Relevant log output

```shell
[testcontainers.org 00:00:00.38] Searching Docker registry credential in CredHelpers
[testcontainers.org 00:00:00.38] Searching Docker registry credential in CredsStore
[testcontainers.org 00:00:00.38] Searching Docker registry credential in Auths
[testcontainers.org 00:00:00.38] Docker registry credential https://index.docker.io/v1/ found
[testcontainers.org 00:00:01.50] Docker image testcontainers/ryuk:0.6.0 created
[testcontainers.org 00:00:01.58] Docker container 8d1b2fa17535 created
[testcontainers.org 00:00:01.64] Start Docker container 8d1b2fa17535
[testcontainers.org 00:00:01.96] Wait for Docker container 8d1b2fa17535 to complete readiness checks
[testcontainers.org 00:00:01.96] Docker container 8d1b2fa17535 ready
[testcontainers.org 00:00:01.97] Searching Docker registry credential in Auths
[testcontainers.org 00:00:01.97] Searching Docker registry credential in Auths
[testcontainers.org 00:00:01.97] Searching Docker registry credential in CredHelpers
[testcontainers.org 00:00:01.97] Searching Docker registry credential in CredsStore
[testcontainers.org 00:00:01.97] Docker registry credential mcr.microsoft.com not found
[testcontainers.org 00:00:18.94] Docker image mcr.microsoft.com/mssql/server:2022-latest created
[testcontainers.org 00:00:18.96] Docker container 4a3b482d21c9 created
[testcontainers.org 00:00:18.97] Start Docker container 4a3b482d21c9
[testcontainers.org 00:00:19.20] Wait for Docker container 4a3b482d21c9 to complete readiness checks
[testcontainers.org 00:00:19.20] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
[testcontainers.org 00:00:20.27] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
[testcontainers.org 00:00:21.32] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
[testcontainers.org 00:00:22.42] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
[testcontainers.org 00:00:23.58] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
[testcontainers.org 00:00:24.69] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
...
[testcontainers.org 00:03:37.79] Execute "/opt/mssql-tools/bin/sqlcmd -Q SELECT 1;" at Docker container 4a3b482d21c9
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by outofrange-consulting**
## What does this PR do?

It adds a check on the mssql-tools path which changed during 2022-CU14 update.

## Why is it important?

Because from the 2022-CU14 version, Microsoft changed the path to the mssql tools, breaking CI for many folks who have been using the 2022-latest image tag.

## Related issues

- Closes #1220

## Follow-ups

I'm not a fan of duplicating the logic between WaitUntil and the MsSqlContainer :'(

### CODE REVIEW

**Review by HofmeisterAn — COMMENTED** [review]
Thank you for preparing the PR. I would like to wait a few more days to see if MS decides whether they will add symlink support (or placing the commands in the base path) or not: https://github.com/microsoft/mssql-docker/issues/892#issuecomment-2253179915.

**Comment by outofrange-consulting** [comment]
Sure no problem :)

File: src/Testcontainers.MsSql/MsSqlBuilder.cs [hunk 1]
```diff
@@ -131,12 +131,20 @@ private MsSqlBuilder WithUsername(string username)
...
-
         /// <inheritdoc />
         public async Task<bool> UntilAsync(IContainer container)
         {
-            var execResult = await container.ExecAsync(_command)
+            var hasMsSql18Tools = await container.ExecAsync(new[] { "[", "-f" , "/opt/mssql-tools18/bin/sqlcmd", "]" })
+                .ConfigureAwait(false);
+     
+            string[] command = [ 
+                hasMsSql18Tools.ExitCode == 0 ? "/opt/mssql-tools18/bin/sqlcmd" : "/opt/mssql-tools/bin/sqlcmd",
```
Comment (ShaunEgan) [inline src/Testcontainers.MsSql/MsSqlBuilder.cs]: Consider moving the paths to named variables.

Comment (outofrange-consulting) [reply src/Testcontainers.MsSql/MsSqlBuilder.cs]: Added them as constants and exposed a new internam method on the container to avoid duplication.

**Comment by kescherCode** [comment]
Even if Microsoft adds symlink support, changes will be neccessary depending on tool version (especially regarding encryption flags).

**Comment by mmalvik** [comment]
> Thank you for preparing the PR. I would like to wait a few more days to see if MS decides whether they will add symlink support (or placing the commands in the base path) or not: [microsoft/mssql-docker#892 (comment)](https://github.com/microsoft/mssql-docker/issues/892#issuecomment-2253179915).

Is this PR still in consideration for being merged or not?

**Comment by HofmeisterAn** [comment]
Yes, but I still hope that Microsoft responds to the issue and decides to address it properly in their image. IIRC, this affects older versions as well. If we need to work around the issue, we will have to fix it for the older versions of MSSQL and different versions of `mssql-tools` too 🙃.

**Comment by HofmeisterAn** [comment]
Since Microsoft is taking their time to address the issue, I think we should move forward with this pull request. Could we adjust the command to determine the `sqlcmd` file path to something like this?

    SQLCMD_PATH=$(find /opt/mssql-tools*/bin/sqlcmd -type f 2>/dev/null | head -n 1)

It should work for all image versions, including new releases. WDYT?

**Comment by HofmeisterAn** [comment]
I've changed the strategy and utilized `find` to find the `sqlcmd` binary to avoid relying on fixed paths. This should support upcoming versions and changes as well. I also need to apply the changes to `ExecScriptAsync(string, CancellationToken)` or find a way to share the approach. I just haven't had the time to finalize it yet. We could probably set `sqlCmdFilePath` only once, similar to what we're doing in [Pulsar](https://github.com/testcontainers/testcontainers-dotnet/blob/fffd384731bd280b9eacd84b65b772ed6c9f7846/src/Testcontainers.Pulsar/PulsarBuilder.cs#L162).

**Review by HofmeisterAn — APPROVED** [review]

**Comment by HofmeisterAn** [comment]
Thanks for the initial PR and your patience 🙏. I think we now have a good approach that should work for new releases (and different MSSQL versions) right out of the box.

**Pull request merged**